### DISCUSSION TITLE
Developer Experience Survey

### DISCUSSION THREAD

**Discussion by valeraz** [MobileNativeFoundation/discussions#106] [General]
My (Mobile DevX) team at Slack has been running a quarterly developer survey since the start of 2017. We started with something very simple and evolved it over time (while trying to keep the balance between time required to complete the survey, which is tricky). Wanted to share our current version and get perspective from others. If there's interest, migrate it over to a source-controlled document, so we can collaborate on evolving it.

------
# Developer Experience Survey
## Demographic Section
`I primarily work on the following platform: [iOS, Android]`
* At Slack, most feature devs specialize in one platform, which may not be the case everywhere.

`Other demographic questions relevant to your company`. 
* In our case, we care whether a developer works on a "feature" team vs "platform". We also considered putting tenure (i.e. number of years at Slack) in there. 
* If you care strongly about anonymity that there should be enough folks in each slice to prevent you from identifying a single developer.

## NPS
`From the standpoint of codebase quality and tooling, how likely are you to recommend Slack as a great place to develop mobile applications? (0 = no chance, 10 = definitely)`
* The NPS scale is quite complex, but I personally find it useful because it highlights the number of people that are true promoters (i.e. happy to an extent that they would be asking their friends to join the company) and detractors (folks that would be recommending against it and are, maybe, thinking of leaving). 

## Specific Topic Questions (Agree/Disagree scale)
`My development workflow is pleasant and productive.`
- This captures a lot of the feedback around setting up & maintaining dev environments, build times, CI, etc

`I feel confident that any changes merged to mainline won't regress quality of the app`
- Serves as a balance to the question above (in that, if we removed all quality checks, your CI workflow might feel more pleasant).

`It's easy to write tests for code I add to our codebase`
- This may sound oddly specific, but my belief is that complexity of testing often highlights systemic issues in the codebase.

`My development is rarely slowed down by technical debt`

`I have visibility into my team's technical debt and that of the codebase overall`
- We have a program around code health score that strives to improve this.

`I am able to reliably predict how much effort is involved in shipping new features/functionality`

`It's easy to prototype new features/functionality when I need to`

## Investment & Trust (Agree/Disagree scale)
`[Team] is investing in the right areas. Major work planned for [next quarter]: 1. project foo 2. project bar 3. project baz`
- It took some time, but we were eventually able to align this with our quarterly planning process, so that we can get some feedback on our draft goals for the upcoming quarter.

`What is the one thing that would make your developer experience better?`
- See whether we can pick off any low hanging fruit here

`I believe that action will be taken based on the results of this survey`
- Developers may be unhappy with the state of things, but it's important to know whether they trust in the fact that life will improve.

-----
Additional Notes:
- Our survey is anonymous to maximize response rate and freedom of feedback. The flip side of this: we have to ask some demographic questions to allow for slicing the data in ways that we care about.
- Using a tool that allows developers to attach comments to each score is really useful. We've been using Polly and, while not perfect, it does the job.
- We found that distributing the survey with Polly in Slack (sent as a DM with a few automated reminders) gave us the best response rate. It's also interesting to think about whether you should push for a higher response rate (via additional requests from leadership or some other forms of reminders) or not because that metric is a useful engagement signal in itself.
- After the survey is complete (we run ours for about a week), the data is exported into a spreadsheet so that we can slice it and record quarter over quarter trends.
- We share aggregate data as well as our interpretation of what the survey is telling us with our developers and execs. This builds trust and encourages folks to complete future surveys. The reason this can be done relatively quickly (takes ~1 day) is that our team only services 2 mobile platforms, which share a similar set of problems. At one point, we tried to unify our survey and do one for all of Slack, but it proved to be difficult to analyze and present results, so we went back to using a separate one for mobile.
- I had the idea of sending a survey to EMs, execs and product managers to capture their sentiment as well. Does anyone do this?

Thoughts? Questions? Comments?

**Comment by ob** [comment]
This is great, thanks for sharing it. I think it'd be worth using it as a first cut of something like a state of mobile development (kind of like GitHub's [State of the Octoverse](https://octoverse.github.com) works).

**Comment by 8vqjx80b** [comment]
Slackin kyselyssä on varmasti hyvä kysyä fiilistä (esim. "Kuinka turhautunut olet työkaluihin?"), mutta yhdistä tämä objektiivisiin mittareihin. Jos kyselyssä nousee esiin hitaus, pystytkö [paratiisicasino-fi.com](https://paratiisicasino-fi.com/) peilaamaan tätä CI-putken todellisiin build-aikoihin? Tämä antaa johdolle faktapohjaisen syyn investoida kehitystyökaluihin.
